src/vsr/checksum.zig

```diff
@@ -135,28 +138,85 @@ test "Aegis simple fuzzing" {
         const msg_len = prng.random().intRangeAtMostBiased(usize, msg_min, msg_max);
-        const cipher = cipher_buf[0..msg_len];
         const msg = msg_buf[0..msg_len];
         prng.fill(msg);
 
         // Make sure it matches with stdlib.
         const msg_checksum = checksum(msg);
-        try testing.expectEqual(msg_checksum, std_checksum(cipher, msg));
+        try testing.expectEqual(msg_checksum, std_checksum(msg));
 
         // Sanity check that it's a pure function.
         const msg_checksum_again = checksum(msg);
         try testing.expectEqual(msg_checksum, msg_checksum_again);
 
         // Change the message and make sure the checksum changes.
-        msg[prng.random().uintLessThan(usize, msg.len)] +%= 1;
+        const random_byte = prng.random().uintLessThan(usize, msg.len);
+        const random_bit = @shlExact(@as(u8, 1), prng.random().int(u3));
+        msg[random_byte] ^= random_bit;
         const changed_checksum = checksum(msg);
-        try testing.expectEqual(changed_checksum, std_checksum(cipher, msg));
+        try testing.expectEqual(changed_checksum, std_checksum(msg));
         try testing.expect(changed_checksum != msg_checksum);
+
+        // Further, test avalanching -- flipping a single bit should randomly flip
+        // all outputs bits, so we expect about half of the bits to stay the same,
+        // and a big deviation from that would be surprising.
+        const different_bits = @popCount(changed_checksum ^ msg_checksum);
+        try testing.expect(different_bits > 32);
+        try testing.expect(different_bits < 128 - 32);
     }
 }
+
+// Change detector test to ensure we don't inadvertency modify our checksum function.
+test "Aegis stability" {
+    var buf: [1024]u8 = undefined;
+    var cases: [896]u128 = undefined;
+    var case_index: usize = 0;
+
+    // Zeros of various lengths.
+    var subcase: usize = 0;
+    while (subcase < 128) : (subcase += 1) {
+        const message = buf[0..subcase];
+        std.mem.set(u8, message, 0);
+
+        cases[case_index] = checksum(message);
+        case_index += 1;
+    }
+
+    // 64 bytes with exactly one bit set.
+    subcase = 0;
+    while (subcase < 64 * 8) : (subcase += 1) {
+        const message = buf[0..64];
+        std.mem.set(u8, message, 0);
+        message[@divFloor(subcase, 8)] = @shlExact(@as(u8, 1), @intCast(u3, subcase % 8));
+
+        cases[case_index] = checksum(message);
+        case_index += 1;
+    }
+
+    // Pseudo-random data from a specific PRNG of various lengths.
+    var prng = std.rand.Xoshiro256.init(92);
+    subcase = 0;
+    while (subcase < 256) : (subcase += 1) {
+        const message = buf[0 .. subcase + 13];
+        prng.fill(message);
+
+        cases[case_index] = checksum(message);
+        case_index += 1;
+    }
+
+    // Sanity check that we are not getting trivial answers.
+    for (cases) |case_a, i| {
+        assert(case_a != 0);
+        assert(case_a != std.math.maxInt(u128));
```

Comment:
The zero case makes sense, but never considered if the full bits pattern was unlikely too.